velox/functions/lib/Re2Functions.cpp

```diff
@@ -26,6 +26,56 @@
 #include "velox/vector/FlatVector.h"
 
 namespace facebook::velox::functions {
+
+std::pair<PatternKind, vector_size_t> determinePatternKind(StringView pattern) {
+  vector_size_t patternLength = pattern.size();
+  vector_size_t i = 0;
+  vector_size_t wildcardStart = -1;
+  vector_size_t fixedPatternStart = -1;
+  vector_size_t anyCharacterWildcardCount = 0;
+  vector_size_t singleCharacterWildcardCount = 0;
+  auto patternStr = pattern.data();
+
+  while (i < patternLength) {
+    if (patternStr[i] == '%' || patternStr[i] == '_') {
+      if (wildcardStart != -1) {
+        return std::make_pair(PatternKind::kGeneric, 0);
+      }
+      wildcardStart = i;
+      while (i < patternLength &&
+             (patternStr[i] == '%' || patternStr[i] == '_')) {
+        singleCharacterWildcardCount += (patternStr[i] == '_');
+        anyCharacterWildcardCount += (patternStr[i] == '%');
+        i++;
+      }
+    } else if (fixedPatternStart == -1) {
+      fixedPatternStart = i;
+      i++;
+    } else {
+      i++;
+    }
+  }
+
+  if (fixedPatternStart == -1) {
+    if (!anyCharacterWildcardCount) {
+      return std::make_pair(
+          PatternKind::kWildcardOne, singleCharacterWildcardCount);
+    }
+    return std::make_pair(
+        PatternKind::kWildcardTwo, singleCharacterWildcardCount);
+  } else if (wildcardStart == -1) {
+    return std::make_pair(PatternKind::kFixed, patternLength);
+  } else if (singleCharacterWildcardCount) {
+    return std::make_pair(PatternKind::kGeneric, 0);
+  } else if (fixedPatternStart < wildcardStart) {
+    return std::make_pair(PatternKind::kPrefix, wildcardStart);
+  } else if (wildcardStart < fixedPatternStart) {
```

Comment:
Won't this always be true?

wildcardStart never equals fixedPatternStart, unless these are -1, but these cases have been checked already.